File: Quicksort.c
#include<stdio.h>
#include<stdlib.h>

void display(int arr[], int size){

    for(int i = 0; i < size; i++){
        printf("%d ", arr[i]);
    }

}

void swap(int arr[], int i, int j){

    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;

}

int partition(int arr[], int l, int r){

    int pivot = arr[r];
    int i = l - 1;
    int j = l;

    for(j; j < r; j++){

        if (arr[j] < pivot){

            i++;
            swap(arr, i, j);

        }
    }

    swap(arr, i + 1, r);

    return i + 1;

}

void quicksort(int arr[], int l, int r){

    if (l < r){

        int pivotIndex = partition(arr, l, r);
        quicksort(arr, l, pivotIndex - 1);
        quicksort(arr, pivotIndex + 1, r);

    }
}

int main(){

    int numbers[6] = {4, 6, 1, 3, 5, 2};

    printf("\nBefore sorting:\n");
    display(numbers, 6);
    
    quicksort(numbers, 0, 5);

    printf("\n\nAfter sorting:\n");
    display(numbers, 6);

    return 0;
}

File: Quicksort.java
import java.util.*;

public class Quicksort {
    
    private static void swap(int[] arr, int i, int j) {
        int tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }

    private static int partition(int[] arr, int l, int r){
        int pivot = arr[r];
        int i = l-1;
        int j;
        for (j = l; j <= r; j++){
            if (arr[j] < pivot){
                i++;
                swap(arr, i, j);
            }
        }
        i++;
        swap(arr, i, r);
        return i;
    }

    private static int[] quicksort(int[] arr, int l, int h){
        if (l > h) return arr;
        int piv = partition(arr, l, h);
        quicksort(arr, l, piv-1);
        quicksort(arr, piv+1, h);
        return arr;
    }

    public static void main(String[] args) {
        System.out.println("QuickSort");
        int[] arr = new int[]{9,3, 8, 2, 5, 1, 7, 4};
        arr = quicksort(arr, 0, arr.length-1);
        System.out.print("Sorted array: "+Arrays.toString(arr));
    }
}